pkg/controller/keycloak/keycloak_controller.go

```diff
@@ -263,3 +281,78 @@ func (r *ReconcileKeycloak) ManageSuccess(instance *v1alpha1.Keycloak, currentSt
 	log.Info("desired cluster state met")
 	return reconcile.Result{RequeueAfter: RequeueDelaySeconds}, nil
 }
+
+func (r *ReconcileKeycloak) removeOrphanedResourcesWithOwnerReference(instance *keycloakv1alpha1.Keycloak) error {
+	config, err := config2.GetConfig()
+	if err != nil {
+		return err
+	}
+
+	// Create Kubernetes Client
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	// Create Client which is working with PartialObjectMetadataList
+	metadataClient, err := metadata.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	// List all resources in cluster
+	// TODO: Investigate how to optimize this, it probably doesn't make sense to go through all resources
+	_, listResources, err := k8sClient.Discovery().ServerGroupsAndResources()
+	for _, resources := range listResources {
```

Comment:
So the main question is how to search for all objects belonging to a specific Operator?

*Before I dig in, I highly recommend reading the [Kubernetes GC Manual](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/).*

As mentioned in the documentation:

> In 1.8, Kubernetes automatically sets the value of ownerReference for objects created or adopted by ReplicationController, ReplicaSet, StatefulSet, DaemonSet, Deployment, Job and CronJob.

This means, those objects will be checked automatically here. We don't need to look at them. What we should search for are Secrets, ConfigMaps, Services, Ingress etc. All objects that may potentially be used by our Operator.

I think it's fine to iterate over those objects, but instead of checking the labels, we can inspect both `metadata.ownerReferences[0].controller` (needs to be true), `metadata.ownerReferences[0].kind` needs to be one of `Keycloak`, `KeycloakBackup`, `KeycloakUser`, `KeycloakRealm` or `KeycloakClient` CR and `metadata.ownerReferences[0].name` needs to be a valid CR name. Here's an output of a hand-written command:

```
kubectl get Secrets -o custom-columns="NAME:metadata.name,OWNING-REFERENCE-KIND:metadata.ownerReferences[0].kind,OWNING-REFERENCE-NAME:metadata.ownerReferences[0].name,CONTROLLER:metadata.ownerReferences[0].controller"

NAME                                OWNING-REFERENCE-KIND   OWNING-REFERENCE-NAME           CONTROLLER
builder-dockercfg-9ghcg             Secret                  builder-token-5c2kl             true
builder-token-5c2kl                 <none>                  <none>                          <none>
builder-token-bdnf5                 <none>                  <none>                          <none>
credential-example-keycloak         Keycloak                example-keycloak                true
default-dockercfg-c8xph             Secret                  default-token-99pbb             true
default-token-99pbb                 <none>                  <none>                          <none>
default-token-rzvzz                 <none>                  <none>                          <none>
deployer-dockercfg-5j2pm            Secret                  deployer-token-lsgv9            true
deployer-token-kw7bk                <none>                  <none>                          <none>
deployer-token-lsgv9                <none>                  <none>                          <none>
keycloak-db-secret                  Keycloak                example-keycloak                true
keycloak-operator-dockercfg-5vbkd   Secret                  keycloak-operator-token-jf7j5   true
keycloak-operator-token-jf7j5       <none>                  <none>                          <none>
keycloak-operator-token-z2wlx       <none>                  <none>                          <none>
rhsso-operator-dockercfg-pxs95      Secret                  rhsso-operator-token-2zkql      true
rhsso-operator-token-2zkql          <none>                  <none>                          <none>
rhsso-operator-token-5d9qr          <none>                  <none>                          <none>
sso-x509-https-secret               Service                 keycloak                        <none>
```

From the list the above, `keycloak-db-secret` meets our needs. The same step needs to be performed for all other objects. 

And once we managed to collect all the objects we need to inspect, we need to compare them with Secondary Resources list. If something is not on that list - it needs to be deleted. 

Does it help?

//cc @pb82 @davidffrench